www/docs/server/validators.md

```diff
@@ -0,0 +1,343 @@
+---
+id: validators
+title: Input & Output Validators
+sidebar_label: Input & Output Validators
+slug: /server/validators
+---
+
+tRPC procedures may define validation logic for their input and/or output. There is first class support for many popular validators, and [integrating a validator](#contributing-your-own-validator-library) which we don't directly support in usually straight-forward.
+
+### Input Validators
+
+By defining an input validator, tRPC can check that a procedure call is correct and return a validation error if not. 
+
+To set up an input validator, use the `procedure.input()` method:
+
+```ts twoslash
+// @target: esnext
+import { initTRPC } from '@trpc/server';
+import { z } from 'zod';
+
+export const t = initTRPC.create();
+const publicProcedure = t.procedure
+
+// ---cut---
+
+export const appRouter = t.router({
+  hello: publicProcedure
+    .input(
+      z.object({
+        name: z.string(),
+      }),
+    )
+    .query((opts) => {
+      const name = opts.input.name;
+      //    ^?
+      return {
+        greeting: `Hello ${opts.input.name}`,
+      };
+    }),
+});
+```
+
+`.input` can also be stacked to build more complex types, which is particularly useful when you want to utilise some common input to a collection of procedures in a [middleware](middlewares).
+
+```ts twoslash
+// @target: esnext
+import { TRPCError, initTRPC } from '@trpc/server';
+import { z } from 'zod';
+
+export const t = initTRPC.create();
+
+// ---cut---
+
+const procedureWhichRejectsOutOfTownFolk = t.procedure
+  .input(z.object({ townName: z.string() }))
+  .use((opts) => {
+    if (opts.input.townName !== 'Pucklechurch') {
+      throw new TRPCError({
+        code: 'FORBIDDEN',
+        message: "We don't take kindly to out-of-town folk",
+      });
+    }
+
+    return opts.next();
+  });
+
+export const appRouter = t.router({
+  hello: procedureWhichRejectsOutOfTownFolk
+    .input(
+      z.object({
+        name: z.string(),
+      }),
+    )
+    .query((opts) => {
+      const input = opts.input;
+      //    ^?
+      return {
+        greeting: `Hello ${input.name}, my friend`,
+      };
+    }),
+});
+```
+
+### Output Validators
+
+By defining an output validator, tRPC can also check that data returned from untrusted sources is correct, or to strip extraneous data. Defining outputs is not always as important as defining inputs, since tRPC gives you automatic type-safety by inferring the return type of your procedures.
+
+Validators are used to infer the input and output types of your procedures.
+
+:::info
+If output validation fails, the server will respond with an `INTERNAL_SERVER_ERROR`.
+:::
+
+## The most basic validator: a function
+
+You can define a validator without any 3rd party dependencies, with a function. This isn't necessarily the best choice to use widely, unless you have a specific need, but it's important to understand that there's no magic here - it's *just typescript*!
```

Comment:
```suggestion
You can define a validator without any 3rd party dependencies, with a function. We don't recommend doing this unless you have a specific need, but it's important to understand that there's no magic here - it's *just typescript*!
```

Maybe also mention the "spec" for validators?
edit: saw the link to the interfaces below, but some reference to what a validator should do (return stuff or throw, anything else?) might make sense here